### PR TITLE
feat: Adding AWS S3 as notification method options for Alerts/Reports

### PR DESCRIPTION
**Pull request by NeHaSinghTeqfocus**
### SUMMARY

We are adding the implementation and usage of the AWS S3 upload feature in Apache Superset. This feature allows users to send reports directly to an AWS S3 bucket. There are three methods available for authentication: taking credentials from the user in the frontend, taking credentials from the superset_config.py file, or utilizing an AWS IAM role associated with the ECS task role.

#### Installation and Setup

- Ensure Apache Superset is installed and running.

- Ensure S3 bucket policy is configured for interacting with AWS S3.

- Configure Apache Superset to enable the AWS S3 upload feature.

- Follow the method-specific setup instructions outlined below.

#### Common Configuration for Enabling AWS S3 feature for all below method.

- User should update the superset_config.py (docker/pythonpath_dev/superset_config.py) file with following:

- Uncomment the following code in superset_config.py.

```
# AWS S3 reporting - uncomment to use
# FEATURE_FLAGS = {"ALERT_REPORTS": True, "ENABLE_AWS": True}
#ALERT_REPORTS_NOTIFICATION_DRY_RUN = False
# AWS Credentials
#AWS_ACCESS_KEY = '####'
#AWS_SECRET_KEY = '####'
```

#### Method 1: Frontend User Credentials

1. When accessing the feature in the frontend, users will be prompted to enter their AWS credentials.

2. Users should provide their AWS Access Key and the AWS Secret Access Key in the user interface itself.

3. Ensure proper IAM policies are applied to the user's credentials to grant necessary permissions for uploading to the specified S3 bucket.

#### Method 2: config.py Credentials

1. Update the config.py file in your Apache Superset installation directory.

2. Add the AWS credentials as configuration variables in superset_config.py.

3. Ensure proper IAM policies are applied to the user's credentials to grant necessary permissions for uploading to the specified S3 bucket.

#### Method 3: AWS IAM Role

1. Ensure that the ECS task running Apache Superset is associated with an IAM role.

2. Configure the IAM role's permissions to allow uploading to the desired S3 bucket.

3. Apache Superset will automatically use the IAM role associated with the ECS task for authentication when uploading to S3.

#### Usage

Once setup has been done, the AWS S3 can be used for sending reports/alert directly to the S3 bucket using following0 AWS S3 method options:

- AWS_S3_credentials

- AWS_S3_pyconfig

- AWS_S3_IAM

· Navigate to settings, then on Alerts & Reports.

· Create alert /report like we use to do with the Email or Slack.

· In the Notification Method select S3.

· Then select the available S3 method options: AWS_S3_credentials, AWS_S3_pyconfig, AWS_S3_IAM.

· Fill the required aws credentials in case of AWS_S3_credentials else just give the bucket name to use.

Like Email, we can also create the reports directly to the Aws S3 bucket from the dashboard and charts page. All three S3 method options would be available from the dashboard and charts page.

· Navigate to any of the dashboard or Chart.

· Create the report using Set up a S3 report and use any of the S3 method options available.

· Generated report can be seen in the Alerts & Reports page.

Note: If the User doesn’t enable the feature flag of ENABLE_AWS, but has ALERT_REPORTS set to True then Aws s3 option wouldn’t be available but he can still use the default Notification method from the Apache Superset i.e. of Email.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ x ] Required feature flags:
- [ x ] Changes UI
- [ x ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ x ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by john-bodley** [comment]
Thanks @NeHaSinghTeqfocus for the PR. It looks like you'll need to rebase your change.

**Comment by geido** [comment]
@rtexelm @fisjac you might want to give this a look

File: superset/reports/notifications/S3.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+
+class S3Notification(BaseNotification):
+
+    type = ReportRecipientType.S3
+    def _get_inline_files(self):
+        current_datetime = datetime.datetime.now()
+        formatted_date = current_datetime.strftime("%Y-%m-%d")
+        report_name= self._content.name
+        name_prefix=f"{report_name}/{formatted_date}/"
```
Comment (craig-rueda) [inline superset/reports/notifications/S3.py:32]: Can we read this from a config prop? Folks might wanna tune the path on their own

File: superset/reports/notifications/S3.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            f'{name_prefix}{report_name}-{str(uuid4())[:8]}.csv': self._content.csv
+            }
+            return data
+        if self._content.screenshots:
+            images = {f'{name_prefix}Screenshot-{str(uuid4())[:8]}.png': screenshot
+                for screenshot in self._content.screenshots}
+            return images
+        return []
+    def _execute_s3_upload(
+            self, file_body, bucket_name, contentType,
```
Comment (craig-rueda) [inline superset/reports/notifications/S3.py:45]: naming: `contentType` -> `content_type`

File: superset/commands/report/execute.py [hunk 3]
```diff
@@ -408,7 +418,13 @@ def _send(
         """
         notification_errors: list[SupersetError] = []
         for recipient in recipients:
-            notification = create_notification(recipient, notification_content)
+            if recipient.type == ReportRecipientType.S3:
+                aws_Configuration = self._get_aws_configuration()
+                notification = create_notification(
+                    recipient, notification_content, aws_Configuration
```
Comment (craig-rueda) [inline superset/commands/report/execute.py:424]: `aws_Configuration` -> `aws_configuration`

File: superset/commands/report/execute.py [hunk 4]
```diff
@@ -396,6 +397,15 @@ def _get_notification_content(self) -> NotificationContent:
             header_data=header_data,
         )
 
+    def _get_aws_configuration(self) -> AwsConfiguration:
+        aws_key = self._report_schedule.aws_key
+        aws_secretKey = self._report_schedule.aws_secretKey
```
Comment (craig-rueda) [inline superset/commands/report/execute.py:402]: aws_secretKey - snake case pls :)

**Review by craig-rueda — CHANGES_REQUESTED** [review]
Thanks for your contribution! Found a few camel->snake case nits. Please also add tests for the backend changes.

**Comment by Ralkion** [comment]
This is being re-created as a separate PR: https://github.com/apache/superset/pull/27594

This one can be closed.

**Comment by Ralkion** [comment]
@eschutho - could you close this PR please?

**Comment by rusackas** [comment]
Closing in favor of https://github.com/apache/superset/pull/27594 but happy to reopen this if @NeHaSinghTeqfocus wants to pick it back up and/or @Ralkion 's doesn't merge for some reason.

**Comment by Ralkion** [comment]
Oh for sure. Thank you @NeHaSinghTeqfocus. They were just unable to continue with the work, so I've picked it up.

**Pull request closed without merging**